### PR TITLE
plan: pick all possible useful indices for merge join.

### PR DESCRIPTION
**Pull request by hanfei1991**
We can get all possible physical properties in advance for smj and stream aggregation. This pr is a pre-version.

### CODE REVIEW

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -213,10 +213,10 @@ func extractTableAlias(p LogicalPlan) *model.CIStr {
 }
 
 func (b *planBuilder) buildJoin(join *ast.Join) LogicalPlan {
-	b.optFlag = b.optFlag | flagPredicatePushDown
 	if join.Right == nil {
 		return b.buildResultSetNode(join.Left)
 	}
+	b.optFlag = b.optFlag | flagPredicatePushDown
```
Comment (shenli) [inline plan/logical_plan_builder.go]: Why move this here?

File: plan/predicate_push_down.go [hunk 2]
```diff
@@ -68,15 +68,17 @@ func (p *LogicalJoin) PredicatePushDown(predicates []expression.Expression) (ret
...
-	groups, valid := tryToGetJoinGroup(p)
-	if valid {
-		e := joinReOrderSolver{allocator: p.allocator, ctx: p.ctx}
-		e.reorderJoin(groups, predicates)
-		newJoin := e.resultJoin
-		parent := p.parents[0]
-		newJoin.SetParents(parent)
-		parent.ReplaceChild(p, newJoin)
-		return newJoin.PredicatePushDown(predicates)
+	if !UseDAGPlanBuilder(p.ctx) { // close join reorder for new plan.
```
Comment (shenli) [inline plan/predicate_push_down.go]: Why close this?

File: plan/property_cols_prune.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	p.rightProperties = rightProperties
+	resultProperties := make([][]*expression.Column, len(leftProperties), len(leftProperties)+len(rightProperties))
+	copy(resultProperties, leftProperties)
+	resultProperties = append(resultProperties, rightProperties...)
+	return resultProperties
+}
+
+func (p *LogicalAggregation) preparePossibleOrderCols() [][]*expression.Column {
+	p.possibleProperties = p.children[0].(LogicalPlan).preparePossibleProperties()
+	return nil
```
Comment (shenli) [inline plan/property_cols_prune.go]: Why return nil?

File: plan/predicate_push_down.go (hunk 2 above)
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Join reorder should be implemented in cbo

File: plan/logical_plan_builder.go (hunk 1 above)
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: It is more reasonable

File: plan/property_cols_prune.go (hunk 3 above)
Comment (hanfei1991) [inline plan/property_cols_prune.go]: Ordered columns cannot be propagated for aggregation

**Comment by shenli** [comment]
LGTM

**Comment by hanfei1991** [comment]
@XuHuaiyu @zimulala PTAL

**Comment by shenli** [comment]
@winoros PTAL

File: plan/new_physical_plan_builder.go [hunk 4]
```diff
@@ -157,9 +157,14 @@ func (p *LogicalJoin) convertToIndexJoin(prop *requiredProp, outerIdx int) (task
 		leftConds     expression.CNFExprs
 		innerTask     task
 		err           error
-		innerJoinKeys = make([]*expression.Column, 0, len(p.EqualConditions))
-		outerJoinKeys = make([]*expression.Column, 0, len(p.EqualConditions))
+		innerJoinKeys []*expression.Column
+		outerJoinKeys []*expression.Column
 	)
+	if outerIdx == 0 {
+		outerJoinKeys, innerJoinKeys = p.LeftJoinKeys, p.RightJoinKeys
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: This can move to line 173.

File: plan/new_physical_plan_builder.go [hunk 5]
```diff
@@ -175,10 +180,6 @@ func (p *LogicalJoin) convertToIndexJoin(prop *requiredProp, outerIdx int) (task
 		switch x := innerChild.(type) {
 		case *DataSource:
 			indices, includeTableScan := availableIndices(x.indexHints, x.tableInfo)
-			for _, cond := range p.EqualConditions {
-				innerJoinKeys = append(innerJoinKeys, cond.GetArgs()[1-outerIdx].(*expression.Column))
-				outerJoinKeys = append(outerJoinKeys, cond.GetArgs()[outerIdx].(*expression.Column))
-			}
 			if includeTableScan {
 				if len(innerJoinKeys) == 1 {
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Why not use "includeTableScan &&  len(innerJoinKeys) == 1" directly?

File: plan/property_cols_prune.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		col := p.getPKIsHandleCol()
+		if col != nil {
+			result = append(result, []*expression.Column{col})
+		}
+	}
+	for _, idx := range indices {
+		var cols []*expression.Column
+		for _, idxCol := range idx.Columns {
+			found := false
+			for _, col := range p.schema.Columns {
```
Comment (zimulala) [inline plan/property_cols_prune.go]: Can we use `schema.FindColum` or `p.findColumn` directly?

Comment (hanfei1991) [inline plan/property_cols_prune.go]: The type of `idxCol` is model.IndexColumn, so we cannot do that.

Comment (zimulala) [inline plan/property_cols_prune.go]: How about "table.FindCol"?

File: plan/property_cols_prune.go (hunk 3 above)
Comment (zimulala) [inline plan/property_cols_prune.go]: Add comments.

File: plan/property_cols_prune.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+func (p *Selection) preparePossibleProperties() (result [][]*expression.Column) {
+	return p.children[0].(LogicalPlan).preparePossibleProperties()
+}
+
+func (p *baseLogicalPlan) preparePossibleProperties() [][]*expression.Column {
+	return nil
+}
+
+func (p *LogicalJoin) preparePossibleOrderCols() [][]*expression.Column {
```
Comment (zimulala) [inline plan/property_cols_prune.go]: Is this name wrong？ It's different from other structures.

File: plan/property_cols_prune.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	// TODO: We should consider properties propagation.
+	p.leftProperties = leftProperties
+	p.rightProperties = rightProperties
+	resultProperties := make([][]*expression.Column, len(leftProperties), len(leftProperties)+len(rightProperties))
+	copy(resultProperties, leftProperties)
+	resultProperties = append(resultProperties, rightProperties...)
+	return resultProperties
+}
+
+func (p *LogicalAggregation) preparePossibleOrderCols() [][]*expression.Column {
```
Comment (zimulala) [inline plan/property_cols_prune.go]: Ditto.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**